src/metabase/models/permissions_v2.clj

```diff
@@ -0,0 +1,231 @@
+;;
+;;  - When fetching a user's permissions, the default behavior is to return the *most* permissive value from any group the
+;;    user is in. This can be overridden by definding a custom implementation of `coalesce`.
+;;
+;;  - If a user does not have any value for the permission when it is fetched, the *least* permissive value is used as a
+;;    fallback.
+
+;; Permissions which apply to individual databases or tables
+(def ^:private DataPermissions
+  {:data-access           {:values [:unrestricted :no-self-service :block]
+                           :model  :model/Table}
+   :native-query-editing  {:values [:yes :no]
+                           :model  :model/Database}
+   :download-results      {:values [:one-million-rows :ten-thousand-rows :no]
+                           :model  :model/Table}
+   :manage-table-metadata {:values [:yes :no]
+                           :model  :model/Table}
+   :manage-database       {:values [:yes :no]
+                           :model  :model/Database}})
+
+;; Permissions which apply to collections
+(def ^:private CollectionPermissions
+  {:collection {:values [:curate :view :no-access]
+                :model  :model/Collection}})
+
+;; Permissions which apply to the application as a whole, rather than being linked to a specific model
+(def ^:private ApplicationPermissions
+  {:settings-access          {:values [:yes :no]}
+   :monitoring-access        {:values [:yes :no]}
+   :subscriptions-and-alerts {:values [:yes :no]}})
+
+(def ^:private Permissions
+  (merge DataPermissions
+         CollectionPermissions
+         ApplicationPermissions))
+
+(def PermissionType
+  "Malli spec for valid permission types."
+  (into [:enum] (keys Permissions)))
+
+
+;;; ------------------------------------------- Misc Utils ------------------------------------------------------------
+
+(defn- least-permissive-value
+  "The *least* permissive value for a given perm type. This value is used as a fallback when a user does not have a
+  value for the permission in the database."
+  [perm-type]
+  (-> Permissions perm-type :values last))
+
+(defn- most-permissive-value
+  "The *most* permissive value for a given perm type. This is the default value for superusers."
+  [perm-type]
+  (-> Permissions perm-type :values first))
+
+(defn- assert-required-object-id
+  "Takes a permission type and a possibly-nil object ID, and throws an exception if the permission type requires an
+  object ID and the object ID is nil."
+  [perm-type object-id]
+  (when (and (get-in Permissions [perm-type :model])
+             (not object-id))
+    (throw (ex-info (tru "Permission type {0} requires an object ID" perm-type)
+                    {perm-type (Permissions perm-type)}))))
+
+(def perm-types-by-model
+  "A map from model identifiers to a list of permission types that apply to that model."
+  (reduce-kv
+   (fn [m perm-type {:keys [model]}]
+     (when model (update m model (fnil conj #{}) perm-type)))
```

Comment:
this should probably be `(if model (update ...) m)` otherwise you risk throwing away your built up state

Reply:
Haha yeah I actually had fixed that but it got reverted when I undid other changes. Good catch

Reply:
nit/wondering out loud: is this `(merge-with (fnil conj #{}) Permissions)` ?

Reply:
nope! but this reducing function would be equivalent:

```
(fn [acc perm-type {:keys [model]}]
     (cond-> acc
       model (update model (fnil conj #{}) perm-type)))
```